### ISSUE TITLE
[IDEA] The th-renaming-tiddler should have a "wiki" argument

### ISSUE THREAD

**Issue by MythosOnGit** [TiddlyWiki/TiddlyWiki5#6536]
Hello everyone. First time posting. Please let me know if I'm not doing it right.

The th-renaming-tiddler hook only takes "newTiddler" and "oldTiddler" as arguments. It doesn't take "wiki", so when that hook is invoked, I have no way of knowing in what wiki the rename is taking place. I can only assume it's $tw.wiki, but that makes writing a testing framework to test my hook is impossible.

https://github.com/Jermolene/TiddlyWiki5/blob/b529e69289ae9027ba032d5782cd752757cd4bbf/core/modules/wiki-bulkops.js#L26

Would it be possible for that hook to take "wiki" as a third argument? Or is there some other way I can determine which wiki I'm working with inside the hook without installing a new hook for every instantiated wiki?

**Comment by pmario** [comment]
Hi, ... You do it right. 

It's possible to have different wiki instances, but at the moment there is only 1. 

We use jasmine as a test framework. At the beginning of this test https://github.com/Jermolene/TiddlyWiki5/blob/master/editions/test/tiddlers/tests/test-wikitext-tabs-macro.js, there is some code that creates a wiki instance and adds some tiddlers. 

$tw.hooks are global at the moment. So that should be enough, to be able to test stuff. 

Usually we add an `options` object at the end of a function call, to extend it in a forward compatible way. But I think that's not needed at the moment.

**Comment by MythosOnGit** [comment]
Thank you for replying.

I'm trying to use Jasmine just like it's used in the tiddlywiki testing framework. But that's the problem. If my test creates a new wiki to test with, then I can't really test a "th-renaming-tiddler" hook which is meant to change something else in the same wiki which contains the renaming tiddlers. Maybe an example will help explain.

```
it('tests', function() {
  var wiki = new $tw.Wiki()
  wiki.addTiddlers(...)
  wiki.renameTiddler('X', 'Y');
  // test here to make sure tiddler 'Z' was updated, but I can't because the rename hook my plugin installed has no way of accessing this local test wiki.
})
```

**Comment by pmario** [comment]
Ah OK, I see what you meant. ... You can have a look at the [tests/data](https://github.com/Jermolene/TiddlyWiki5/tree/master/editions/test/tiddlers/tests/data). It contains some tiddlers that are automatically included in the "test main wiki". So those tiddlers can be accessed using. $tw.wiki.getTiddler() ... 

So you don't need the `new $tw.Wiki()` initialisation. You can add your tiddlers in a new directory. .. That isn't ideal, but it should work.

**Comment by MythosOnGit** [comment]
You're right. That really is not ideal. I can't isolate my tests very well if they're all using the $tw.wiki.

It really seems like the "options" argument like you mentioned before would fit right in with a hook like that, but if there's no chance of changing the hook, then I'll trying something like this:
```
it('tests', function() {
  var wiki = new $tw.Wiki()
  var oldWiki = $tw.wiki;
  try {
    $tw.wiki = wiki;
    wiki.addTiddlers(...)
    wiki.renameTiddler('X', 'Y');
    // test to ensure 'Z' was updated
  } finally {
    $tw.wiki = oldWiki;
  }
})
```

Should I close this issue?

**Comment by pmario** [comment]
That seems to be an "OK" work around. .. @Jermolene .. what do you think?

**Comment by MythosOnGit** [comment]
It doesn't actually work well as a workaround. If any module accessing or "require" statements occur in that "try" block, then your tests crash.

**Comment by pmario** [comment]
@MythosOnGit ... Did you ever solve the problem?
